### PR TITLE
[Testing] Refactored and tested data processing in mrmc_method.py

### PR DESCRIPTION
**Pull request by jakeval**
Previously, processing data for recourse was handled clumsily between MRM and MRMC and made it difficult to get the correct order of operations between processing, filtering data points, and clustering. Processing and filtering data points has been altered and unit tests have been added to mrmc_method.py.

## Changelist
* filter_data() has been removed and is now part of MRM.process_data()
* Optional kwargs have been added to MRM and MRMC for adding confidence thresholds
* The private __init__ argument _processed_data was added to MRM so that it doesn't re-process data when initialized within MRMC
* unit tests were added to cover the changes

## Feedback requested
* Are the changes to filtering and data processing clear?
* Are the unit tests readable?
* Is there sufficient unit test coverage in mrmc_method.py?

### CODE REVIEW

File: recourse_methods/mrmc_method.py [hunk 1]
```diff
@@ -92,47 +92,85 @@ def alpha(dist: np.ndarray) -> np.ndarray:
...
+        rescale_direction: Optional[RecourseRescaler] = normalize_rescaler,
+        confidence_threshold: Optional[float] = None,
+        model: Optional[model_interface.Model] = None,
+        _processed_data: Optional[recourse_adapter.EmbeddedDataFrame] = None,
     ):
         """Creates an MRM instance.
 
+        The dataset is processed for recourse using MRM.process_data(). This
+        processing step can be skipped by providing recourse data directly via
+        _processed_data. This is typically unnecessary for end users.
```
Comment (bringingjoy) [inline recourse_methods/mrmc_method.py:111]: Nit: describing the function `MRM.process_data()` and private class attribute `_processed_data` is a little confusing. I'd suggest changing the 2nd sentence to:

"The data processing step can be skipped by directly assigning the class attribute `_processed_data` with data ready for recourse."

or something like this.

File: recourse_methods/mrmc_method.py [hunk 2]
```diff
@@ -92,47 +92,85 @@ def alpha(dist: np.ndarray) -> np.ndarray:
 
 
 class MRM:
```
Comment (bringingjoy) [inline recourse_methods/mrmc_method.py:94]: I'd recommend following this class docstring example as described here: https://google.github.io/styleguide/pyguide.html#384-classes

(Move the class initialization details to just below the `class MRM:` definition and out of the `__init__()` method's docstring.)

File: recourse_methods/mrmc_method.py [hunk 3]
```diff
@@ -92,47 +92,85 @@ def alpha(dist: np.ndarray) -> np.ndarray:
...
         dataset: pd.DataFrame,
         adapter: recourse_adapter.RecourseAdapter,
+        confidence_threshold: Optional[float] = None,
+        model: Optional[model_interface.Model] = None,
     ) -> recourse_adapter.EmbeddedDataFrame:
         """Processes the dataset for MRM.
-        It filters out negative outcomes and transforms the data to a numeric
-        embedded space.
+
+        It filters out negative outcome points, drops the class label, and
```
Comment (bringingjoy) [inline recourse_methods/mrmc_method.py:153]: Suggestion for explicitness and clarity: "Removes datapoints with negative (-1) labels, drops the class label feature column, ... Also excludes datapoints with below-threshold model confidence."

File: recourse_methods/mrmc_method.py [hunk 4]
```diff
@@ -178,7 +197,7 @@ def get_unnormalized_direction(
         diff = (self.data - poi).to_numpy()
         dist = np.sqrt(np.power(diff, 2).sum(axis=1))
         alpha_val = self.alpha(dist)
-        if alpha_val.isnull().any():
+        if np.isnan(alpha_val).any():
             raise RuntimeError(
                 f"Alpha function returned null values: {alpha_val}"
```
Comment (bringingjoy) [inline recourse_methods/mrmc_method.py:202]: In RuntimeError: "returned NaN values"

File: recourse_methods/mrmc_method.py [hunk 5]
```diff
@@ -337,8 +346,9 @@ def cluster_data(
             cluster_centers=cluster_centers,
         )
 
-    def validate_cluster_assignments(
-        self, cluster_assignments: pd.DataFrame, k_directions: int
+    @staticmethod
+    def _validate_cluster_assignments(
+        cluster_assignments: pd.DataFrame, k_directions: int
     ) -> bool:
         """Raises an error if the cluster_assignments is valid.
```
Comment (bringingjoy) [inline recourse_methods/mrmc_method.py:353]: I think this should be "raises an error if the cluster_assignments are _invalid_".

File: recourse_methods/mrmc_method.py [hunk 6]
```diff
@@ -263,56 +283,45 @@ def __init__(
...
-            return self
-
-    # TODO(@jakeval): Confidence check this
-    def cluster_data(
-        self, data: recourse_adapter.EmbeddedDataFrame, k_directions: int
+    @staticmethod
+    def _cluster_data(
+        data: recourse_adapter.EmbeddedDataFrame, k_directions: int
     ) -> Clusters:
         """Clusters the data using k-means clustering.
```
Comment (bringingjoy) [inline recourse_methods/mrmc_method.py:326]: I'd add using sklearn's KMeans clustering.

**Review by bringingjoy — APPROVED** [review]
Thanks, jakeval@! Some minor comments.

File: recourse_methods/mrmc_method.py (hunk 2 above)
Comment (jakeval) [reply recourse_methods/mrmc_method.py:94]: I fixed this in this file and added an [issue](https://github.com/jakeval/MRMC/issues/34) to review the rest of the repo.

**Pull request merged**